Account-History v0.3

maelsar opened the pull request:
closes #56 

Feature for MVP from issue #1

- [x] Current orders list
- [x] Orders history list
- [x] Trade history list
- [ ] Cancel unfilled limit order / partially filled

TODO enhancements:
- [ ] change pair onClick from the table
- [ ] Show all orders function
- [ ] Export CSV function
- [ ] Make Buttons "Show all orders and export as CSV" only displayable depending on active tables
- [ ] Rerender history when wallet is disconnected
NOTE - working when initially not-connected to connected but history should disappear from connected to disconnected
- [ ] Rerender history after a transaction has been made
- [ ] Add datafield for avg executed price when available from Adex
- [ ] Display "Connect Wallet to see History" if no wallet is connected
- [ ] Display "No Orders" if Wallet is connected but no History

Other Observations that needs fixing:
- Wallet Balance (front end) does not auto update after a successful transaction.

Next I will focus more on the Cancel Order rather than the enhancements.

EvgeniiaVak commented:
@maelsar this PR description contains many unchecked boxes, are they for future PRs? Or do you wish to continue updating this one and we shouldn't merge it yet (if so, you can convert this PR to draft, so that people wouldn't be able merge yet)?

maelsar commented:
@EvgeniiaVak the unchecked boxes will be for future PRs and can be classified under enhancements.  For this PR the base feature is already working other than the "Cancel Order".  I think this might help others as well to visualize the testing the orders moving forward.

maelsar commented:
@EvgeniiaVak if this PR is acceptable then perhaps this could close this issue #56 then create another issue for:
Cancel Orders and another for the mentioned enhancements

EvgeniiaVak commented:
@maelsar ok, please do cut and paste all the undone features from here into one or more new issues, because once this PR is merged, #56 will be closed

EvgeniiaVak commented on src/app/RenderTable.tsx:
We may need separate components btw, because the columns in "Trade History" and "Order History" are quite different.

maelsar replied:
> Maybe renaming this component to something, RenderTable sounds too generic?.. I'm not sure how to name this component though, it contains "Open Orders", "Order History", and "Trade History", correct?

To be honest I was also confused how to name this component.  I would gladly accept any suggestions from everyone to find the better replacement.

> We may need separate components btw, because the columns in "Trade History" and "Order History" are quite different.

OrderReceipts response gives all the orders from the account. Currently to display the "Order History" I am displaying unfiltered orders (COMPLETED/PENDING/CANCELED) while "Trade History" filters only the orders that have been labeled as COMPLETE.
While OrderReceipts also gives all the "trades" that make up to complete one order, i opted not to include that as it might be too much information for the average user.

I am open to suggestions on how to differentiate between "Order History" and "Trade History"

Suggestion: perhaps we could also change "Open Orders" to "Active Orders".  Maybe it would make more sense to the users?

fliebenberg replied:
I would suggest we have "Open Orders" that include only orders with status "PENDING". We then also have "Order History" which shows all the orders (status PENDING/CANCELLED/COMPLETED). We can also show "User Trades" which are the individual trades for the user's orders, but I agree that might be too complicated for most users and not necessary for launch.

maelsar replied:
> I would suggest we have "Open Orders" that include only orders with status "PENDING". We then also have "Order History" which shows all the orders (status PENDING/CANCELLED/COMPLETED). We can also show "User Trades" which are the individual trades for the user's orders, but I agree that might be too complicated for most users and not necessary for launch.

Yes, this is currently the case for "Open Orders". It is filtering for Orders with status "PENDING".  

Regarding the "Trade History": I think the fields [Pair--Direction--Qty.--Avg. Filled(or Executed) Price--Order Fee--Completed Time] with filter for "COMPLETED".
As a trader myself, I think these are the fields I look at when checking my history.  What do you guys think?
@fliebenberg will you be including the "Avg. Filled(or Executed) Price" in the OrderReceipt or should I handle it on the frontend? Also, would it be possible to include the Order Fee to be included OrderReceipt?

EvgeniiaVak replied:
I think it's not very complicated and even a beginner trader understands the difference between orders and trades. As for what columns to include, I think this discussion is better be done in figma, so that @steferr could updated the wireframes appropriately. Also we can start removing the extra and debugging and stuff from the page, and only have the content specified in the wireframe.

EvgeniiaVak commented on src/app/RenderTable.tsx:
```diff
@@ -0,0 +1,100 @@
+        <table>
+          <thead>
+            <tr>
+              <th>Pair</th>
+              <th>ID</th>
```
What are these columns? I think we need only human readable Pair Name, no need for pair ID, if it's order id it should be later in the columns
please refer to figma to get the correct columns seprarately for all 3 tables: - https://www.figma.com/file/P7pfzKwJ4G6ClapXfl61D3/Dexter-UI?type=design&node-id=352-20276&mode=design&t=WbFm5oHBrdWQTz3h-0

maelsar replied:
`<th>Pair</th>` would display the pair name ie ADEX / XRD etc
`<th>ID</th>` while i dont think this is relevant information for the average user, i just included it to be displayed for now just to display, but i can remove that so the average user wouldnt fixate on what that is.

Will update this to be closer to the figma display.

fliebenberg replied:
I agree that any of the id fields are probably not relevant for end-users so would agree that we remove all ids from the ui. I do understand it is very useful when developing though.

maelsar commented on src/app/RenderTable.tsx:
```diff
@@ -0,0 +1,100 @@
+  selectedTable: string | null;
+}
+
+function formatDate(date: string) {
```
will do

maelsar commented on src/app/RenderTable.tsx:
```diff
@@ -0,0 +1,100 @@
+}
+
+export function RenderTable({ response, selectedTable }: RenderTableProps) {
+  const adexState = useContext(AdexStateContext);
```
`const adexState = useContext(AdexStateContext);` your right, i think i overlooked this. will remove.

fliebenberg commented:
> @EvgeniiaVak the unchecked boxes will be for future PRs and can be classified under enhancements. For this PR the base feature is already working other than the "Cancel Order". I think this might help others as well to visualize testing the orders moving forward.

Just a reminder that the cancel order functionality in the SDK works now so it should be as easy as providing a cancel button and connecting it to the submitCancelTx function in the Alphadex SDK.

maelsar commented:
> > @EvgeniiaVak the unchecked boxes will be for future PRs and can be classified under enhancements. For this PR the base feature is already working other than the "Cancel Order". I think this might help others as well to visualize testing the orders moving forward.
> 
> Just a reminder that the cancel order functionality in the SDK works now so it should be as easy as providing a cancel button and connecting it to the submitCancelTx function in the Alphadex SDK.

I already have a working cancel order function in another branch albeit still needs a bit more testing and some refactoring.  After I make the changes, I will probably make a new PR for the CancelOrder.

EvgeniiaVak commented:
> I already have a working cancel order function in another branch albeit still needs a bit more testing and some refactoring. After I make the changes, I will probably make a new PR for the CancelOrder.

@maelsar I would be glad to see the cancel feature in this PR as well, but whatever is easier for your workflow.

maelsar commented:
> @maelsar I would be glad to see the cancel feature in this PR as well, but whatever is easier for your workflow.

Noted. will add this to feature to this PR.

maelsar commented:
I addressed your request for changes @EvgeniiaVak 
- Added the Cancel Order and should be working.
Note:
There is still a placeholder for column [Order Price] as this will be added in the future as discussed with @fliebenberg 
I will also add all the individual trades of a particular order and that will also be added in OrderReceipts as discussed with Fred
![3](https://github.com/DeXter-on-Radix/website/assets/107124432/ef5f67f4-6194-4611-a24d-480cd6f457d1)
![1](https://github.com/DeXter-on-Radix/website/assets/107124432/b68398df-c535-457f-a8c9-17f9e21e4953)
![2](https://github.com/DeXter-on-Radix/website/assets/107124432/470d8170-32de-4253-bf9f-bae98d1c4a6e)

maelsar commented:
Other Notes:
I just changed RenderTable to DisplayTable for now.
I will still refactor DisplayTable in the future when we can finalize the data columns we decide to show, for now i kept it that way to be easier to manipulate the table layout in the meantime.

EvgeniiaVak commented:
I get an error when running npm run dev:
TypeError: Cannot read properties of undefined (reading 'filter')
![Screenshot 2023-08-15 at 13 57 33](https://github.com/DeXter-on-Radix/website/assets/27793901/a1d4a7c6-62bf-4774-bbd4-2c6909937a7a)

EvgeniiaVak commented:
@maelsar could you please use types (not `any[]`)? I'll wire up the linter to complain about it, the point of typescript is to use types to catch bugs at compile time, if we just silence types we will not have this benefit.

EvgeniiaVak commented on src/app/page.tsx:
```diff
@@ -24,11 +26,23 @@ export default function Home() {
     ) : null;
   }
 
+  function getAccountHistory() {
+    //gets Orderlist
+    //TODO:
+    //Rerender this component when wallet is disconnected
+    //Rerender when a transaction has been made
+    const account = wallet ? wallet.accounts[0]?.address : "";
+    return adexState.status != "LOADING" ? (
+      <AccountHistory account={account} />
+    ) : null;
+  }
+
```
I don't think we really want to show nothing when we have no data. We've done so with other components in the past, but that's not good - will cause layout shifts and page jumps when stuff loads. Better to show some empty rows.

maelsar replied:
This should still render a the table headers but with no row. Will check again and test to at least have 1 empty row.

EvgeniiaVak commented:
About the component name, maybe "AccountHistory"?

EvgeniiaVak commented:
> I get an error when running npm run dev: TypeError: Cannot read properties of undefined (reading 'filter') ![Screenshot 2023-08-15 at 13 57 33](https://user-images.githubusercontent.com/27793901/260682024-a1d4a7c6-62bf-4774-bbd4-2c6909937a7a.png)

That's maybe due to network errors today (e.g. I still get 500 https://ansharnet-gateway.radixdlt.com/state/entity/details in console log on other branches), but we need to check for those (or null values resulting from those) and not crash our page also.

maelsar commented:
> About the component name, maybe "AccountHistory"?

My parent component is already named AccountHistory, DisplayTable is the child component.

maelsar commented:
I will work on the other changes you mentioned.  I also get the gateway error right now.

maelsar commented:
Main concerns: 
1. Component now renders even with a gateway error 500 (see attached picture)
![5](https://github.com/DeXter-on-Radix/website/assets/107124432/4a6949b8-2934-4af0-82c7-05b9760ded62)

2. code now doesn't use type any [ ]

maelsar commented:
Made a conditional display of rows if
1. No wallet is connected
2. wallet is connected but no history data
Notes: 
- Table will still render with table headers even if no data is available to prevent layout shift.
- DisplayTable will still be refactored in a later PR after finalization of data to be displayed.

see photos for tests 
![active 1](https://github.com/DeXter-on-Radix/website/assets/107124432/e16f1c4f-2448-4593-b372-8c7c7e6be7aa)
![active 2](https://github.com/DeXter-on-Radix/website/assets/107124432/5fc3c9f3-3f60-4152-8c38-f852fc1b0a94)
![active 3](https://github.com/DeXter-on-Radix/website/assets/107124432/7cd0c60b-5aef-4932-a015-aeaa98151509)
![cancelled](https://github.com/DeXter-on-Radix/website/assets/107124432/47a35f7f-b107-4929-abad-7613ca2a250c)
![connected no active 1](https://github.com/DeXter-on-Radix/website/assets/107124432/f6a0a818-67c9-4f1f-8e03-420065e210e0)
![no hist - 1](https://github.com/DeXter-on-Radix/website/assets/107124432/723f2cb8-c7b5-4185-b373-0c2096d45b65)
![no hist - 2](https://github.com/DeXter-on-Radix/website/assets/107124432/165fc7aa-e588-4455-a450-d07497fdae9c)
![no hist - 3](https://github.com/DeXter-on-Radix/website/assets/107124432/1389b16d-a335-48e5-ace5-8f922cd9452b)
![no wallet - 1](https://github.com/DeXter-on-Radix/website/assets/107124432/2e2e3330-b051-468e-b51c-3b3047eac8b7)
![no wallet - 2](https://github.com/DeXter-on-Radix/website/assets/107124432/653d145d-e650-485a-89d4-597818d7cba7)
![no wallet - 3](https://github.com/DeXter-on-Radix/website/assets/107124432/57bda7ed-7db5-4864-aa4b-7c2a56ba21c4)

EvgeniiaVak reviewed (changes requested):
@maelsar please update to the latest main, let me know if you need my help with updating to redux

maelsar commented:
Will let you know if I need help on this. I mentioned earlier that I had to take some time off and will return on Tuesday (tomorrow) to work on this. Thamks!

maelsar commented:
This PR can be closed. It was easier for me to rebuild as a new branch.

EvgeniiaVak commented:
#89
